### PR TITLE
Gcc11 gcc12 + conan v2

### PR DESCRIPTION
**Pull request by boofhead**
Specify library name and version:  gcc/11.3.0 and gcc/12.1.0

Add latest versions of gcc

This is also a good place to share with all of us **why you are submitting this PR** (specially if it is a new addition to ConanCenter): is it a dependency of other libraries you want to package? Are you the author of the library? Thanks!

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by AndreyMlashkin** [comment]
@boofhead are you able to use a recipe? If so, how?
I couldn't use cross compilation with gcc/10.2.0

**Comment by boofhead** [comment]
> @boofhead are you able to use a recipe? If so, how?
> I couldn't use cross compilation with gcc/10.2.0

Hi Andrey,

Looks like cross building is not supported. Apart form that it works fine for me in a Conan virtualenv as is.

```
    def configure(self):
        if self.settings.os == "Windows":
            raise ConanInvalidConfiguration("Windows builds aren't supported (yet), sorry")
        if tools.cross_building(self.settings):
            raise ConanInvalidConfiguration("no cross-building support (yet), sorry")
```

**Comment by AndreyMlashkin** [comment]
Okey, I haven't tried to use virtual env. I have only used this gcc recipe in build_requirements and it was not working for me.

**Comment by boofhead** [comment]
I think setting the compiler in build requires is sort of a chicken and egg problem. I couldn't work out how/if it was supposed to work.

**Comment by AndreyMlashkin** [comment]
@boofhead, could you please cherry-pick my commit?
https://github.com/conan-io/conan-center-index/pull/12259/commits/3b9a23d8b4e189592b6af1da4f25f8e96fa58a0d

**Comment by boofhead** [comment]
Hi Andrey, Sorry, I don't know how to cherry-pick a commit from your fork. Are there instructions somewhere you could point me at?

**Comment by AndreyMlashkin** [comment]
You can simply copy commit's content then :)

**Comment by AndreyMlashkin** [comment]
It seems, that only a recipe layout should be corrected a bit

File: recipes/gcc/all/conandata.yml [hunk 1]
```diff
@@ -2,3 +2,9 @@ sources:
   "10.2.0":
     sha256: 27e879dccc639cd7b0cc08ed575c1669492579529b53c9ff27b0b96265fa867d
     url: https://ftp.gnu.org/gnu/gcc/gcc-10.2.0/gcc-10.2.0.tar.gz
+  "11.3.0":
```
Comment (prince-chrismc) [inline recipes/gcc/all/conandata.yml:5]: newer can go on top please 🙏

File: recipes/gcc/all/conanfile.py [hunk 2]
```diff
@@ -1,126 +1,112 @@
-from conans import ConanFile, tools, AutoToolsBuildEnvironment
-from conans.errors import ConanException, ConanInvalidConfiguration
 import os
 
-required_conan_version = ">=1.29.1"
+from conan import ConanFile
+from conan.errors import ConanInvalidConfiguration
+from conan.tools.build import cross_building
+from conan.tools.files import get, replace_in_file, rmdir, rm, copy
+from conan.tools.gnu import AutotoolsToolchain, Autotools
```
Comment (prince-chrismc) [inline recipes/gcc/all/conanfile.py:7]: Please add "+ conan v2" in the title !!!! this is amazing

File: recipes/gcc/all/conanfile.py [hunk 3]
```diff
@@ -1,126 +1,112 @@
...
+    topics = 'gcc', 'gnu', 'compiler', 'c', 'c++'
+    homepage = 'https://gcc.gnu.org'
+    url = 'https://github.com/conan-io/conan-center-index'
+    license = 'GPL-3.0-only'
+    settings = 'os', 'compiler', 'arch', 'build_type'
+    requires = 'mpc/1.2.0', 'mpfr/4.1.0', 'gmp/6.2.1', 'zlib/1.2.12', 'isl/0.24'
+    tool_requires = 'flex/2.6.4'
+    short_paths = True
+
+    _auto_tools = None
```
Comment (prince-chrismc) [inline recipes/gcc/all/conanfile.py:27]: This should no longer be cached with the new ones

File: recipes/gcc/all/conanfile.py [hunk 4]
```diff
@@ -1,126 +1,112 @@
...
-        extracted_dir = "gcc-%s" % self.version
-        os.rename(extracted_dir, self._source_subfolder)
+        get(self, **self.conan_data["sources"][self.version], strip_root=True)
 
     @property
-    def _make_args(self):
-        if self.settings.os == "Macos":
-            return ["BOOT_LDFLAGS=-Wl,-headerpad_max_install_names"]
-        return []
+    def auto_tools(self):
```
Comment (prince-chrismc) [inline recipes/gcc/all/conanfile.py:66]: not required anymore

File: recipes/gcc/all/conanfile.py [hunk 5]
```diff
@@ -1,126 +1,112 @@
...
+        rm(self, '*.la', self.package_folder, recursive=True)
+        copy(self, pattern='COPYING*', dst=os.path.join(self.package_folder, 'licenses'), src=self.source_folder)
 
     def package_info(self):
-        bindir = os.path.join(self.package_folder, "bin")
-        self.output.info("Appending PATH env var with : " + bindir)
-        self.env_info.PATH.append(bindir)
+        bindir = os.path.join(self.package_folder, 'bin')
+        self.output.info('Appending PATH env var with : ' + bindir)
+        self.env_info.PATH = [bindir] + self.env_info.PATH
```
Comment (prince-chrismc) [inline recipes/gcc/all/conanfile.py:101]: This is unsual... what exactly is it for?

File: recipes/gcc/all/test_package/conanfile.py [hunk 6]
```diff
@@ -21,7 +24,7 @@ def chmod_plus_x(name):
         self.run("%s -dumpversion" % cxx, run_environment=True)
         self.run("%s %s -o hello_c" % (cc, hello_c), run_environment=True)
         self.run("%s %s -o hello_cpp" % (cxx, hello_cpp), run_environment=True)
-        if not tools.cross_building(self.settings):
+        if not cross_building(self):
```
Comment (prince-chrismc) [inline recipes/gcc/all/test_package/conanfile.py:27]: there s a better "can_run" that can be used here

**Review by prince-chrismc — COMMENTED** [review]
You are also missing the `test_v1_package` ... in the docs there are more details :)

File: recipes/gcc/all/conandata.yml (hunk 1 above)
Comment (boofhead) [reply recipes/gcc/all/conandata.yml:5]: Can do but other packages I looked at are in version order aren't they?

File: recipes/gcc/all/conanfile.py (hunk 2 above)
Comment (boofhead) [reply recipes/gcc/all/conanfile.py:7]: done

File: recipes/gcc/all/conanfile.py (hunk 3 above)
Comment (boofhead) [reply recipes/gcc/all/conanfile.py]: thanks

File: recipes/gcc/all/conanfile.py (hunk 5 above)
Comment (boofhead) [reply recipes/gcc/all/conanfile.py]: made this more Conan v2 style. I think bin dir needs to be on the path if you actually want to build anything with gcc.

**Comment by boofhead** [comment]
> You are also missing the `test_v1_package` ... in the docs there are more details :)

I can't find this is in the docs, can you post a link here?

**Comment by boofhead** [comment]
@prince-chrismc do you understand the default layout hook failure I'm seeing? I don't get this error locally.

File: recipes/gcc/all/conandata.yml (hunk 1 above)
Comment (prince-chrismc) [reply recipes/gcc/all/conandata.yml:5]: https://github.com/conan-io/conan-center-index/blob/4bf114eefecf85163341d07233fa94b6186ae97c/docs/package_templates/cmake_package/config.yml#L2

They should be ordered but just the opposite direction  😅

File: recipes/gcc/all/conanfile.py (hunk 5 above)
Comment (prince-chrismc) [reply recipes/gcc/all/conanfile.py]: Where did you find the "2.0 style"?

As a c++ dev this is way less readable. I had to Google what this does 🙊 the old append was more obvious but maybe I need to work on my python skill

**Comment by AndreyMlashkin** [comment]
> @prince-chrismc do you understand the default layout hook failure I'm seeing? I don't get this error locally.
> 

you need to install conan hooks to get the same warnings.
Most probably, you just need to call rm for home folder in package method

**Comment by prince-chrismc** [comment]
Sorry for missing your question I must have missed the notification 🙈 hopefully this will unblocks next week 🤞

**Comment by boofhead** [comment]
> > @prince-chrismc do you understand the default layout hook failure I'm seeing? I don't get this error locally.
> 
> you need to install conan hooks to get the same warnings. Most probably, you just need to call rm for home folder in package method

@AndreyMlashkin I have the hooks installed locally and I don't get this error. Also this error seems weird. Why would there be a home directory in the package directory. Output of the HOOKS running locally is below

```
[HOOK - conan-center.py] post_package(): [PACKAGE LICENSE (KB-H012)] OK
[HOOK - conan-center.py] post_package(): [DEFAULT PACKAGE LAYOUT (KB-H013)] OK
[HOOK - conan-center.py] post_package(): [MATCHING CONFIGURATION (KB-H014)] OK
[HOOK - conan-center.py] post_package(): [SHARED ARTIFACTS (KB-H015)] OK
[HOOK - conan-center.py] post_package(): [PC-FILES (KB-H020)] OK
[HOOK - conan-center.py] post_package(): [CMAKE-MODULES-CONFIG-FILES (KB-H016)] OK
[HOOK - conan-center.py] post_package(): [PDB FILES NOT ALLOWED (KB-H017)] OK
[HOOK - conan-center.py] post_package(): [LIBTOOL FILES PRESENCE (KB-H018)] OK
[HOOK - conan-center.py] post_package(): [MS RUNTIME FILES (KB-H021)] OK
[HOOK - conan-center.py] post_package(): [SHORT_PATHS USAGE (KB-H066)] OK
/usr/bin/objdump: ./lib/libgcc_s.so: file format not recognized
[HOOK - conan-center.py] post_package(): WARN: [MISSING SYSTEM LIBS (KB-H043)] Running objdump on './lib/libgcc_s.so' failed. Is the environment variable OBJDUMP correctly configured?
[HOOK - conan-center.py] post_package(): [MISSING SYSTEM LIBS (KB-H043)] OK
```

**Comment by AndreyMlashkin** [comment]
Maybe I am blind, but I don't see an actual error message in the logs...

**Comment by AndreyMlashkin** [comment]
If I build from your branch, it fails in test package for me:

conan create . gcc/11.3.0@andrei/test  --build missing -u -r conancenter
conan profile:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.version=11.2
compiler.cppstd=17
compiler.libcxx=libstdc++11
os=Linux
os_build=Linux
[options]
vulkan-loader:shared=True
[build_requires]
[env]



```
...
gcc/11.3.0@andrei/test: Package 'e22f47f196f50cff15626bfbdc57fadd84b7f986' created
gcc/11.3.0@andrei/test: Created package revision 157e5c0a2a76935e9522d0cb97a4648d
gcc/11.3.0@andrei/test: Appending PATH env var with : /home/andrei/.conan/data/gcc/11.3.0/andrei/test/package/e22f47f196f50cff15626bfbdc57fadd84b7f986/bin
gcc/11.3.0@andrei/test: Creating CC env var with : /home/andrei/.conan/data/gcc/11.3.0/andrei/test/package/e22f47f196f50cff15626bfbdc57fadd84b7f986/bin/gcc-11.3.0
gcc/11.3.0@andrei/test: Creating CXX env var with : /home/andrei/.conan/data/gcc/11.3.0/andrei/test/package/e22f47f196f50cff15626bfbdc57fadd84b7f986/bin/g++-11.3.0
gcc/11.3.0@andrei/test (test package): Generator txt created conanbuildinfo.txt
gcc/11.3.0@andrei/test (test package): Generator cmake created conanbuildinfo.cmake
gcc/11.3.0@andrei/test (test package): Aggregating env generators
gcc/11.3.0@andrei/test (test package): Generated conaninfo.txt
gcc/11.3.0@andrei/test (test package): Generated graphinfo
Using lockfile: '/home/andrei/sources/conan-center-index/recipes/gcc/all/test_package/build/05471bae34a664fe3f352cf69c545297c1aaf924/conan.lock'
Using cached profile from lockfile
gcc/11.3.0@andrei/test (test package): Calling build()
gcc/11.3.0@andrei/test (test package): WARN: This conanfile has no build step
gcc/11.3.0@andrei/test (test package): Running test()
ERROR: gcc/11.3.0@andrei/test (test package): Error in test() method, line 17
	cc = os.environ["CC"]
	KeyError: 'CC'
```

**Comment by boofhead** [comment]
> If I build from your branch, it fails in test package for me:
> 
> conan create . gcc/11.3.0@andrei/test --build missing -u -r conancenter conan profile: [settings] arch=x86_64 arch_build=x86_64 build_type=Release compiler=gcc compiler.version=11.2 compiler.cppstd=17 compiler.libcxx=libstdc++11 os=Linux os_build=Linux [options] vulkan-loader:shared=True [build_requires] [env]
> 
> ```
> ...
> gcc/11.3.0@andrei/test: Package 'e22f47f196f50cff15626bfbdc57fadd84b7f986' created
> gcc/11.3.0@andrei/test: Created package revision 157e5c0a2a76935e9522d0cb97a4648d
> gcc/11.3.0@andrei/test: Appending PATH env var with : /home/andrei/.conan/data/gcc/11.3.0/andrei/test/package/e22f47f196f50cff15626bfbdc57fadd84b7f986/bin
> gcc/11.3.0@andrei/test: Creating CC env var with : /home/andrei/.conan/data/gcc/11.3.0/andrei/test/package/e22f47f196f50cff15626bfbdc57fadd84b7f986/bin/gcc-11.3.0
> gcc/11.3.0@andrei/test: Creating CXX env var with : /home/andrei/.conan/data/gcc/11.3.0/andrei/test/package/e22f47f196f50cff15626bfbdc57fadd84b7f986/bin/g++-11.3.0
> gcc/11.3.0@andrei/test (test package): Generator txt created conanbuildinfo.txt
> gcc/11.3.0@andrei/test (test package): Generator cmake created conanbuildinfo.cmake
> gcc/11.3.0@andrei/test (test package): Aggregating env generators
> gcc/11.3.0@andrei/test (test package): Generated conaninfo.txt
> gcc/11.3.0@andrei/test (test package): Generated graphinfo
> Using lockfile: '/home/andrei/sources/conan-center-index/recipes/gcc/all/test_package/build/05471bae34a664fe3f352cf69c545297c1aaf924/conan.lock'
> Using cached profile from lockfile
> gcc/11.3.0@andrei/test (test package): Calling build()
> gcc/11.3.0@andrei/test (test package): WARN: This conanfile has no build step
> gcc/11.3.0@andrei/test (test package): Running test()
> ERROR: gcc/11.3.0@andrei/test (test package): Error in test() method, line 17
> 	cc = os.environ["CC"]
> 	KeyError: 'CC'
> ```

Will see if I can reproduce it when I get a chance but that looks pretty strange

> gcc/11.3.0@andrei/test: Creating CC env var with : /home/andrei/.conan/data/gcc/11.3.0/andrei/test/package/e22f47f196f50cff15626bfbdc57fadd84b7f986/bin/gcc-11.3.0

> ERROR: gcc/11.3.0@andrei/test (test package): Error in test() method, line 17
> 	cc = os.environ["CC"]
> 	KeyError: 'CC'

**Pull request closed without merging**